### ISSUE TITLE
"okta_app_oauth / pkce_required" attribute required if "token_endpoint_auth_method" is "none"

### ISSUE THREAD

**Issue by tmatilai** [okta/terraform-provider-okta#1325]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
erraform v1.3.1
on darwin_amd64
+ provider registry.terraform.io/okta/okta v3.37.0
```

### Affected Resource(s)

- `okta_app_oauth`

### Terraform Configuration Files

```hcl
resource "okta_app_oauth" "test" {
  label = "test"
  type  = "browser"

  grant_types    = ["authorization_code", "implicit"]
  response_types = ["code", "id_token", "token"]
  redirect_uris  = ["http://example.com/"]

  token_endpoint_auth_method = "none"
}
```

### Output

<details><summary>When upgrading from <code>&lt; v3.36</code> provider version</summary>

```
$ terraform apply

okta_app_oauth.test: Refreshing state... [id=0oa4klsz1mRcelZWg0x7]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # okta_app_oauth.test will be updated in-place
  ~ resource "okta_app_oauth" "test" {
        id                         = "0oa4klsz1mRcelZWg0x7"
        name                       = "oidc_client"
      - pkce_required              = true -> null
        # (29 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

okta_app_oauth.test: Modifying... [id=0oa4klsz1mRcelZWg0x7]
╷
│ Error: failed to update OAuth application: the API returned an error: Api validation failed: App Instance. Causes: errorSummary: ''pkce_required'' must be set to true when ''token_endpoint_auth_method'' is ''none''.
│
│   with okta_app_oauth.test,
│   on okta_app.tf line 1, in resource "okta_app_oauth" "test":
│    1: resource "okta_app_oauth" "test" {
│
╵
```
</details>

<details><summary>When creating new resource with <code>&gt;= v3.36</code> provider version</summary>

```
$ terraform apply

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # okta_app_oauth.test will be created
  + resource "okta_app_oauth" "test" {
      + accessibility_self_service = false
      + auto_key_rotation          = true
      + auto_submit_toolbar        = false
      + client_id                  = (known after apply)
      + client_secret              = (sensitive value)
      + consent_method             = "TRUSTED"
      + grant_types                = [
          + "authorization_code",
          + "implicit",
        ]
      + hide_ios                   = true
      + hide_web                   = true
      + id                         = (known after apply)
      + issuer_mode                = "ORG_URL"
      + label                      = "test"
      + login_mode                 = "DISABLED"
      + logo_url                   = (known after apply)
      + name                       = (known after apply)
      + omit_secret                = false
      + redirect_uris              = [
          + "http://example.com/",
        ]
      + refresh_token_leeway       = (known after apply)
      + refresh_token_rotation     = (known after apply)
      + response_types             = [
          + "code",
          + "id_token",
          + "token",
        ]
      + sign_on_mode               = (known after apply)
      + skip_groups                = false
      + skip_users                 = false
      + status                     = "ACTIVE"
      + token_endpoint_auth_method = "none"
      + type                       = "browser"
      + user_name_template         = "${source.login}"
      + user_name_template_type    = "BUILT_IN"
      + wildcard_redirect          = "DISABLED"
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

okta_app_oauth.test: Creating...
╷
│ Error: failed to create OAuth application: the API returned an error: Api validation failed: App Instance. Causes: errorSummary: ''pkce_required'' must be set to true when ''token_endpoint_auth_method'' is ''none''.
│
│   with okta_app_oauth.test,
│   on okta_app.tf line 1, in resource "okta_app_oauth" "test":
│    1: resource "okta_app_oauth" "test" {
│
╵
```
</details>

### Expected Behavior

Based on [the documentation](https://registry.terraform.io/providers/okta/okta/latest/docs/resources/app_oauth#pkce_required), the `pkce_required` should have been `true` by default for `browser` and `native` app types.

So either the functionality or documentation should be modified.

Having to specify the attribute breaks compatibility with older provider versions. This is especially unwanted in child modules, as some root modules might need to pin down the provider version for some other bugs.

### Actual Behavior

Okta API errors out.

### Steps to Reproduce

1. `terraform apply`

### References

- #1305

**Comment by monde** [comment]
@tmatilai I'll see if I can do a quick fix and release on this bug since it was just introduced.

**Comment by tmatilai** [comment]
Oh, and even without the error, this probably causes drift in the state, and Terraform trying to always remove the attribute.

**Comment by monde** [comment]
Thanks @tmatilai . I have a fix for you in #1327 but I'm not sure if I'll get released Friday 10/07/2022 as I'd like to take a stab at fixing #655 also. This solve was a bit bedeviling for me given the TF SDK behavior around booleans.

**Comment by tmatilai** [comment]
Thanks a lot @monde!

A quick test seems to confirm that #1327 works as expected. No rush with the release, we can always pin down the provider version.

**Comment by monde** [comment]
fix released as v3.38.0 https://github.com/okta/terraform-provider-okta/releases/tag/v3.38.0
